### PR TITLE
Introduce fallback procedure

### LINKED ISSUE #938: Enable fallback procedure in rusk node
#### Summary
 This effort is about migrating current (golang) implementation of fallback procedure into rust. As we've already introduced `InSync and OutOfSync` chain states, another one `InFallback` can be used to facilitate the fallback procedure. Below is a list of requirements to be taken into account.  

- detect securely an event of a consensus split
- allow  node state transition from `InSync` into `InFallback`
- disallow node state transition from `OutOfSync` into `InFallback`
- trigger a procedure of fallback 
- trace all steps

If feasible, a unit-test/script that simulates a consensus split may be provided as well.

### PR DESCRIPTION
**Pull request by goshawk-3**
fixes #938

### CODE REVIEW

File: node/src/chain/fallback.rs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+                    /// block
+                    warn!("deleting a block from first iteration");
+                }
+
+                Ledger::delete_block(t, &chain_blk)?;
+
+                // Attempt to resubmit transactions back to mempool.
+                // An error here is not considered critical.
+                for tx in blk.txs().iter().skip(1) {
+                    // TODO: Filter out Distribute tx explicitly
```
Comment (herr-seppia) [inline node/src/chain/fallback.rs:160]: I don't think we still have `distribute` txs

File: node/src/chain/acceptor.rs [hunk 2]
```diff
@@ -117,6 +119,31 @@ impl<DB: database::DB, VM: vm::VMExecution, N: Network> Acceptor<N, DB, VM> {
         })
     }
 
+    /// Injects a block in ledger database without updating VM state.
+    ///
+    /// It also updates mrb and provisioners list accordingly.
+    ///
+    /// Injecting block is applicable only in case of the fallback procedure
+    /// execution.
```
[the remark this answers is not in the document]
Comment (goshawk-3) [reply node/src/chain/acceptor.rs:127]: The fallback procedure starts with a call to `VM.Revert()`. Inject_block is actually updating ledger state to be aligned with the already updated VM state.

File: node/src/chain/fallback.rs (hunk 1 above)
Comment (goshawk-3) [reply node/src/chain/fallback.rs:160]: Good catch!

File: node/src/chain/acceptor.rs [hunk 3]
```diff
@@ -117,6 +119,47 @@ impl<DB: database::DB, VM: vm::VMExecution, N: Network> Acceptor<N, DB, VM> {
...
+        }
+
+        // Reset Consensus
+        task.abort_with_wait().await;
+
+        //  Update register.
+        self.db
+            .read()
+            .await
+            .update(|t| t.set_register(&blk.header))?;
```
Comment (herr-seppia) [inline node/src/chain/acceptor.rs:155]: It's not related to this PR, but having a `lock.read()` that can perform `write` operations is quit weird

I know that rocksdb doesn't require a `mut`, but maybe we have to force it

File: node/src/chain/acceptor.rs [hunk 4]
```diff
@@ -125,17 +168,24 @@ impl<DB: database::DB, VM: vm::VMExecution, N: Network> Acceptor<N, DB, VM> {
...
+            &blk.header,
+        )
+        .await?;
+
         // Reset Consensus
         task.abort_with_wait().await;
 
         // Persist block in consistency with the VM state update
-        let updated_provisioners = {
+        {
```
Comment (herr-seppia) [inline node/src/chain/acceptor.rs:188]: Any specific reason to move the update of the provisioners inside this scope?

**Review by herr-seppia — APPROVED** [review]
LGTM

**Review by HDauven — APPROVED** [review]

File: node/src/chain/acceptor.rs (hunk 3 above)
Comment (goshawk-3) [reply node/src/chain/acceptor.rs:155]: DB component implements interior mutability and rocksdb member is thread-safe. If we use `self.db.write()`  here it means that any concurrent read query will be blocked.

A few examples of concurrent read queries that we should not block:
- Block explorer requesting ledger data
- Responding to GetData/GetBlocks/GetTx kadcast messages on sync-up procedure.

File: node/src/chain/acceptor.rs (hunk 4 above)
Comment (goshawk-3) [reply node/src/chain/acceptor.rs:188]: Yes, for consistency with updating the mrb field.

**Pull request merged**